### ISSUE TITLE
How to broadcast training graph from master node

### ISSUE THREAD

**Issue by tremblerz** [uber/horovod#355]
I have been trying to run distributed training using tensorflow for a case where worker nodes don't have the `training_graph` but the master node does. There is no success so far with tensorflow distributed training, I wanted to know if horovod has some api to make it possible. Going by this [doc](https://www.tensorflow.org/versions/master/deploy/distributed), it seems like every node has to construct the graph on its own and only later they can synchronize it, but in my case only master node has the training graph in the beginning and I want it to broadcast this graph to worker nodes or worker nodes to pull the graph somehow.

Please see [this question](https://stackoverflow.com/questions/51146348/how-to-pass-computation-graph-or-train-op-in-distributed-tensorflow) if I haven't clarified the problem sufficiently.

**Comment by alsrgv** [comment]
@tremblerz, Horovod has no facility for doing this, but one simple way could be to save checkpoint on the shared filesystem and restore from it on all the workers.  If you don't have shared filesystem, you can load the file as a byte array and broadcast it.  If you're running master w/o MPI, this would require #129.  Please let us know if this route sounds attractive.

**Comment by tremblerz** [comment]
Thanks @alsrgv !
Broadcasting the file as byte array suits my requirement well. If I haven't got it wrong then cycle would like this ->
`master creates and checkpoints the metagraph -> master broadcast this metagraph -> workers load the metagraph -> workers run the training op`
right?

**Comment by alsrgv** [comment]
@tremblerz, that's right.  Are you running master & workers together in a single MPI job?  Does master have a GPU?

**Comment by alsrgv** [comment]
@tremblerz, in order to use Horovod for training you need to run MPI job among the homogeneous GPU workers.  It will be possible to have a separate master that is not part of MPI job, does not have a GPU and just prepares the graph when #129 is implemented.  You'd still need some mechanism to transmit graph file to workers (MPI broadcast will not work since master will not be part of an MPI job) and kick off `mpirun` job among that will read the graph and start the training.  Would that work for you?

**Comment by tremblerz** [comment]
@alsrgv Thanks for clarifying. That sounds good enough. Will give it a shot and let you know.

**Comment by alsrgv** [comment]
@tremblerz, sounds good.  You can install from #357 to pick the change that allows creating the graph in non-MPI environment and then running it on MPI workers.

**Comment by vitalyli** [comment]
What is driving this requirement, just curious?

**Comment by alsrgv** [comment]
@tremblerz, were you able to make use of #357?

**Comment by alsrgv** [comment]
@tremblerz, did you make progress with your use case?

**Comment by tremblerz** [comment]
Hi @alsrgv sorry for the late reply. I got stuck with some other issues in the normal training. I will get back on it and let you know if I happen to use the described method.
